rust: add support for Amba device drivers.

wedsonaf opened the pull request:
This is used by the PL061 gpio driver.

Based on #544

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

bjorn3 commented on rust/kernel/amba.rs:
```diff
@@ -0,0 +1,268 @@
+    /// Device id.
+    pub id: u32,
+
+    /// Mask that identifies which bits are valid in the device d.
```
*device id

wedsonaf replied:
Fixed.

bjorn3 commented on rust/kernel/amba.rs:
```diff
@@ -0,0 +1,268 @@
+        ];
+    };
+
+    // Conver case without a trailing comma.
```
*convert

wedsonaf replied:
Fixed.

ojeda commented on rust/kernel/amba.rs:
```diff
@@ -0,0 +1,268 @@
+/// A registration of an amba driver.
+pub type Registration<T> = driver::Registration<Adapter<T>>;
+
+/// ID of an Amba device.
```
Nit: we are using `id`, `ID`, etc. Perhaps we could use `ID` for "prose" and

    `id`

(i.e. with backquotes) to refer to arguments/members/etc.?

wedsonaf replied:
Now that I think about, I don't think I like "ID" because it's not an acronym. I replaced all usages with just lower case "id", except when it's the first word of a sentence, in which case I use "Id".

wedsonaf commented:
PTAL.

### v1 -> v2

* Fixed typos in comments
* Fixed usage of "ID" and "id" in comments
* Brought back `#![allow(dead_code)]` in `io_mem` and `power` because Amba are arm-specific, so the warnings still appeared in other archs. I'll remove them again when I add a subsystem that is arch-agnostic.
* Made publishing `module_amba_driver` in `prelude` only when `CONFIG_ARM_AMBA` is set.
* Updated `module_amba_driver` to use the new syntax of the first arg of `module_driver`
